drivers/interrupt_controller/intc_uwp.c

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2018, UNISOC Incorporated
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <soc.h>
+#include <errno.h>
+#include <assert.h>
+#include <irq_nextlevel.h>
+
+#include "intc_uwp.h"
+#include "uwp_hal.h"
+
+/* convenience defines */
+#define DEV_DATA(dev) \
+	((struct intc_uwp_data * const)(dev)->driver_data)
+#define INTC_STRUCT(dev) \
+	((volatile struct uwp_intc *)(DEV_DATA(dev))->base_addr)
+
+struct intc_uwp_data {
+	u32_t	base_addr;
+	u32_t	irq_table_offset;
+};
+
+static ALWAYS_INLINE void uwp_dispatch_child_isrs(u32_t intr_status,
+						      u32_t isr_base_offset)
+{
+	u32_t intr_bitpos, intr_offset;
+
+	/* Dispatch lower level ISRs depending upon the bit set */
+	while (intr_status) {
+		intr_bitpos = find_lsb_set(intr_status) - 1;
+		intr_status &= ~(1 << intr_bitpos);
+		intr_offset = isr_base_offset + intr_bitpos;
+		_sw_isr_table[intr_offset].isr(
+			_sw_isr_table[intr_offset].arg);
+	}
+}
+
+static void intc_uwp_irq(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	struct intc_uwp_data *intc_data = DEV_DATA(dev);
+	volatile struct uwp_intc *intc = INTC_STRUCT(dev);
+
+	uwp_dispatch_child_isrs(uwp_intc_status(intc),
+			intc_data->irq_table_offset);
+}
+
+static inline void uwp_intc_irq_enable(struct device *dev, unsigned int irq)
+{
+	volatile struct uwp_intc *intc = INTC_STRUCT(dev);
+
+	uwp_intc_enable(intc, irq);
+}
+
+static inline void uwp_intc_irq_disable(struct device *dev, unsigned int irq)
+{
+	volatile struct uwp_intc *intc = INTC_STRUCT(dev);
+
+	uwp_intc_disable(intc, irq);
+}
+
+static inline u32_t uwp_intc_irq_get_state(struct device *dev)
+{
+	volatile struct uwp_intc *intc = INTC_STRUCT(dev);
+
+	return uwp_intc_status(intc);
+}
+
+void uwp_intc_irq_trigger_soft(struct device *dev)
+{
+	volatile struct uwp_intc *intc = INTC_STRUCT(dev);
+
+	uwp_intc_trigger_soft(intc);
+}
+
+void uwp_intc_irq_clear_soft(struct device *dev)
+{
+	volatile struct uwp_intc *intc = INTC_STRUCT(dev);
+
+	uwp_intc_clear_soft(intc);
+}
+
+static const struct irq_next_level_api uwp_intc_apis = {
+	.intr_enable = uwp_intc_irq_enable,
+	.intr_disable = uwp_intc_irq_disable,
+	.intr_get_state = uwp_intc_irq_get_state,
+};
+
+#ifdef CONFIG_INTC_IRQ_UWP
+static struct device DEVICE_NAME_GET(intc_irq_uwp);
+static struct intc_uwp_data intc_irq_dev_data = {
```

Comment:
Please follow the same structure followed in other interrupt drivers. i.e a config structure which is a constant and another non-const structure. Something like below.
static const struct cavs_ictl_config cavs_config_1 = {
        .irq_num = CAVS_ICTL_1_IRQ,
         .isr_table_offset = CONFIG_CAVS_ISR_TBL_OFFSET +
                             CONFIG_MAX_IRQ_PER_AGGREGATOR,
         .config_func = cavs_config_1_irq,
};

static struct cavs_ictl_runtime cavs_1_runtime = {
         .base_addr = CAVS_ICTL_BASE_ADDR + sizeof(struct cavs_registers),
};